HyperVExtension/src/HyperVExtension/Services/VMGalleryService.cs

```diff
@@ -0,0 +1,129 @@
+        AllowTrailingCommas = true,
+    };
+
+    public VMGalleryService(IDownloaderService downloaderService)
+    {
+        _downloaderService = downloaderService;
+    }
+
+    private static readonly Uri VmGalleryUrl = new("https://go.microsoft.com/fwlink/?linkid=851584");
+
+    private VMGalleryImageList _imageList = new();
+
+    /// <inheritdoc cref="IVMGalleryService.GetGalleryImagesAsync"/>
+    public async Task<VMGalleryImageList> GetGalleryImagesAsync()
+    {
+        // If we have already downloaded the image list before, return it.
+        if (_imageList.Images.Count > 0)
+        {
+            return _imageList;
+        }
+
+        var emptyList = new VMGalleryImageList();
+        try
+        {
+            var cancellationTokenSource = new CancellationTokenSource();
+
+            // This should be quick as the file is around 28.0 KB, so a 5 minute timeout should be ok on the worst network.
+            cancellationTokenSource.CancelAfter(TimeSpan.FromMinutes(5));
+
+            // get the JSON data
+            var resultJson = await _downloaderService.DownloadStringAsync(VmGalleryUrl.AbsoluteUri, cancellationTokenSource.Token);
+            _imageList = JsonSerializer.Deserialize(resultJson, typeof(VMGalleryImageList), _jsonOptions) as VMGalleryImageList ?? emptyList;
+
+            // Now we need to download the base64 images for the symbols (icons). So they can be used within an adaptive card.
+            foreach (var image in _imageList.Images)
+            {
+                if (!string.IsNullOrEmpty(image.Symbol.Uri))
+                {
+                    var byteArray = await _downloaderService.DownloadByteArrayAsync(image.Symbol.Uri, cancellationTokenSource.Token);
+                    if (!ValidateFileSha256Hash(byteArray, image.Symbol.Hash))
+                    {
+                        Logging.Logger()?.ReportError(ComponentName, $"Symbol Hash '{image.Symbol.Hash}' validation failed for image with name '{image}'. Symbol uri: '{image.Symbol}'");
+                        continue;
+                    }
+
+                    image.Symbol.Base64Image = Convert.ToBase64String(byteArray);
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            Logging.Logger()?.ReportError(ComponentName, $"Unable to retrieve VM gallery images", ex);
+        }
+
+        return _imageList;
+    }
+
+    /// <inheritdoc cref="IVMGalleryService.GetDownloadedArchiveFileName"/>
+    public string GetDownloadedArchiveFileName(VMGalleryImage image)
+    {
+        var hash = image.Disk.Hash;
+        var hexValues = hash.Split(':').Last();
+
+        // Now get the file extension from the web uri. The web uri is the location of the zipped image on the web.
+        // we can expect this last segment to be the file name plus the extension.
+        var webUri = new Uri(image.Disk.Uri);
+        var fileExtension = webUri.Segments.Last().Split('.').Last();
```

Comment:
Does `Path.GetExtension()` not work for `URI`?

Reply:
nope only `string` and `ReadOnlySpan<char>`